Presence sample

moozzyk opened the pull request:
Adding a sample of how to handle presence in single server and scaleout scenarios

moozzyk commented:
:up::date: - cleaned up

moozzyk commented:
:watch:

BrennanConroy commented on samples/ChatSample/DefaultPresenceManager.cs:
```diff
@@ -0,0 +1,83 @@
+
+        public async Task UserLeft(Connection connection)
+        {
+            usersOnline.TryRemove(connection, out UserDetails user);
```
Wrap this in an if statement and only execute Notify if true

moozzyk replied:
Fixed.

mikaelm12 commented on samples/ChatSample/DefaultPresenceManager.cs:
```diff
@@ -0,0 +1,84 @@
+            _logger = loggerFactory.CreateLogger<DefaultPresenceManager<THub>>();
+        }
+
+        private readonly ConcurrentDictionary<Connection, UserDetails> usersOnline
```
rename to `_usersOnline`

mikaelm12 replied:
rename to `_usersOnline`

davidfowl commented on samples/ChatSample/DefaultPresenceManager.cs:
```diff
@@ -0,0 +1,84 @@
+        {
+            foreach (var connection in usersOnline.Keys)
+            {
+                using (var scope = _serviceScopeFactory.CreateScope())
```
This is the grossest part of this change. Creating a new hub instance here feels kinda wrong as it requires mimicking the HubEndPoint logic... At least it's a sample and we get to see what's required to make it work.

moozzyk replied:
Yeah. There is not a good way to invoke a hub method from outside.

davidfowl commented on src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs:
```diff
@@ -32,9 +32,14 @@ public HubRouteBuilder(SocketRouteBuilder routes)
         }
 
-        public void MapHub<THub>(string path) where THub : Hub
+        public void MapHub<THub>(string path) where THub : Hub<IClientProxy>
```
why?

moozzyk replied:
I think it was related to the other change. Seems no longer needed when I dropped the TClient.

moozzyk replied:
This is actually needed if you try to map a hub the derives from `Hub<IClientProxy>` instead of `Hub`.  I have `HubWithPresence : HubWithPresence<IClientProxy>` and `HubWithPresence<TClient> : Hub<TClient>` and it is causing an eror: `There is no implicit reference conversion from 'ChatSample.Hubs.Chat' to 'Microsoft.AspNetCore.SignalR.Hub'.` I can create `HubWithPresence : Hub` and it will work but the other should work as well.

davidfowl commented on src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs:
```diff
@@ -32,9 +32,14 @@ public HubRouteBuilder(SocketRouteBuilder routes)
             _routes.MapEndpoint<HubEndPoint<THub>>(path);
         }
+
+        public void MapHub<THub, TClient>(string path) where THub : Hub<TClient>
```
Isn't this overload broken?

moozzyk replied:
I started with `DefaultPresenceManager<THub>` was `DefaultPresenceManager<THub, TClient>` which required this but this indeed did not work (I think I hit: https://github.com/aspnet/SignalR/issues/48).  Fixing https://github.com/aspnet/SignalR/issues/48 did not seem straightforward and was not the main point of this PR. I will remove this overload for now.

davidfowl commented:
I think you should make derived a `HubLifetimeManager` that implements the `IPresenceManager` since all of the state is already there. That should remove the need to pass things in the `HubWithPresence` ctor. We also need to figure out how to help people distinguish between UserOnline and OnConnected (same for disconnected, offline)

davidfowl commented on samples/ChatSample/RedisPresenceManager.cs:
```diff
@@ -0,0 +1,175 @@
+
+        // TODO: subscribe and handle lifecycle events/disconnects
+        // TODO: handle situations where Redis shuts down
+        // TODO: handle situations where a server goes down and the server has zombie connections
```
This is key to the entire feature

moozzyk replied:
This is not part of this PR.

moozzyk commented on samples/ChatSample/RedisPresenceManager.cs:
```diff
@@ -0,0 +1,175 @@
+
+            _redisSubscriber.Subscribe(_redisChannel, (channel, value) =>
+            {
+                var user = ToUserDetails(((string)value).Substring(1));
```
We can. It's not a cast btw. it is a conversion operator so I don't even know how expensive it is.

moozzyk commented on samples/ChatSample/Views/Home/Index.cshtml:
```diff
@@ -34,6 +64,14 @@ document.getElementById('sendmessage').addEventListener('submit', event => {
     event.preventDefault();
 });
+
+let addUserOnline = function (user) {
```
It was introduced by the person who originally submitted the PR and I did not notice it. I don't think it needs to be a variable at all.

BrennanConroy commented on samples/ChatSample/DefaultPresenceManager.cs:
```diff
@@ -0,0 +1,84 @@
+
+        public async Task UserLeft(Connection connection)
+        {
+            if (usersOnline.TryRemove(connection, out UserDetails user))
```
out var

BrennanConroy commented on samples/ChatSample/RedisPresenceManager.cs:
```diff
@@ -0,0 +1,175 @@
+            connections.TryRemove(connection, out object _);
+
+            var database = _redisConnection.GetDatabase(_redisDatabase);
+            var user = $"{connection.ConnectionId }|{ connection.User.Identity.Name}";
```
format

moozzyk commented:
Rebased and addressed most of the feeback. Working on _make derived a HubLifetimeManager that implements the IPresenceManager_ which I will send separately.

moozzyk commented:
I accidentally delete the branch which closed the PR. I opened a new, follow up PR here: https://github.com/aspnet/SignalR/pull/391
